tests/unit/package_managers/test_pip.py

```diff
@@ -3384,20 +3387,38 @@ def test_download_dependencies(
             "requirement_file": str(req_file.file_path.subpath_from_root),
         }
 
-        sdist = pip.DistributionPackageInfo("foo", "1.0", "sdist", pypi_download, "", False)
-
-        w1_path = pip_deps.join_within_root("foo-1.0-cp25-win32.whl").path
-        w2_path = pip_deps.join_within_root("foo-1.0-cp35-many-linux.whl").path
-        w3_path = pip_deps.join_within_root("foo-1.0-any.whl").path
-
-        wheels = []
+        sdist_DPI = pip.DistributionPackageInfo(
+            "foo",
+            "1.0",
+            "sdist",
+            sdist_download,
+            "",
+            False,
+            pypi_checksums={ChecksumInfo("sha256", "abcdef")},
+        )
+        sdist_d_i = sdist_DPI.download_info | {
+            "kind": "pypi",
+            "requirement_file": str(req_file.file_path.subpath_from_root),
+            "hash_verified": True,
+        }
+        wheels_DPI = []
+        wheels_d_i = []
         if allow_binary:
-            wheels = [
+            wheels_DPI = [
```

Comment:
This whole hunk is too verbose, would it be possible to shorten it e.g. like this (haven't tried the tests though)?
```
-        wheels_d_i = []
         if allow_binary:
-            wheels_DPI = [
-                pip.DistributionPackageInfo(
+            for wheel_path, checksums, hash_verified in zip(
+                [wheel_0_download, wheel_1_download, wheel_2_download],
+                [{ChecksumInfo("sha256", "fedcba")}, {ChecksumInfo("sha256", "abcdef")}, set()],
+                [False, True, False],
+            ):
+                dpi = pip.DistributionPackageInfo(
                     "foo",
                     "1.0",
                     "wheel",
-                    wheel_0_download,
+                    wheel_path,
                     "",
                     False,
-                    pypi_checksums={ChecksumInfo("sha256", "fedcba")},
-                ),
-                pip.DistributionPackageInfo(
-                    "foo",
-                    "1.0",
-                    "wheel",
-                    wheel_1_download,
-                    "",
-                    False,
-                    pypi_checksums={ChecksumInfo("sha256", "abcdef")},
-                ),
-                pip.DistributionPackageInfo(
-                    "foo",
-                    "1.0",
-                    "wheel",
-                    wheel_2_download,
-                    "",
-                    False,
-                    pypi_checksums=set(),
-                ),
-            ]
-            wheels_d_i = [
-                wheels_DPI[0].download_info
-                | {
-                    "kind": "pypi",
-                    "requirement_file": str(req_file.file_path.subpath_from_root),
-                    "hash_verified": False,
-                },
-                wheels_DPI[1].download_info
-                | {
-                    "kind": "pypi",
-                    "requirement_file": str(req_file.file_path.subpath_from_root),
-                    "hash_verified": True,
-                },
-                wheels_DPI[2].download_info
-                | {
-                    "kind": "pypi",
-                    "requirement_file": str(req_file.file_path.subpath_from_root),
-                    "hash_verified": False,
-                },
-            ]
-
-        pypi_downloads = [sdist_d_i] + wheels_d_i
+                    pypi_checksums=checksums,  # type: ignore
+                )
+                wheels_DPI.append(dpi)
+                pypi_downloads.append(
+                    dpi.download_info
+                    | {
+                        "kind": "pypi",
+                        "requirement_file": str(req_file.file_path.subpath_from_root),
+                        "hash_verified": False,
+                    }
+                )
```
IOW instead of explicitly creating each DPI and download_info we'd generate it since 98% of data is identical.

Reply:
I'll try it out

Reply:
I think we reached the point where you can squash the marked commits to their intended places in the next revision, I think we'll deal with it wrt review just fine.

Reply:
Tests fixed